xc/device.ml

```diff
@@ -607,13 +607,36 @@ let qemu_media_change ~xs device _type params =
 
 	if is_upstream_qemu device.frontend.domid
 	then begin
+		let open Qmp in
 		let cd = "ide1-cd1" in
-		let qmp_cmd =
-			if params = ""
-			then (Qmp.Command(None, Qmp.Eject (cd, Some true)))
-			else (Qmp.Command(None, Qmp.Change (cd, params, None))) in
-		qmp_write device.frontend.domid qmp_cmd
-	end
+		match params with
+		| "" ->
+			let qmp_cmd = Command(None, Eject (cd, Some true)) in
+			qmp_write device.frontend.domid qmp_cmd
+		| _ ->
+			let c = Qmp_protocol.connect (Printf.sprintf "/var/run/xen/qmp-libxl-%d" device.frontend.domid) in
+			finally
+				(fun () ->
+					try
+						let buf = " " in
+						let fd_of_c = Qmp_protocol.to_fd c in
+						let fd_of_cd = Unix.openfile params [ Unix.O_RDONLY ] 0o640 in
+						ignore(Fd_send_recv.send_fd fd_of_c buf 0 1 [] fd_of_cd);
+
+						let qmp_cmd = Command (None, Add_fd (Fd_send_recv.int_of_fd fd_of_cd)) in
+						Qmp_protocol.negotiate c;
+						Qmp_protocol.write c qmp_cmd;
+						let qmp_ret = Qmp_protocol.read c in
+						let new_fd = match qmp_ret with
+						| Success (None, Fd_info fd_info) -> fd_info.fd
+						| _ -> raise (Internal_error (Printf.sprintf "Get unexpected result after sending Qmp message: %s" (string_of_message qmp_cmd)))
```

Comment:
@mg12 
Thanks for your comments.
QEMU accepts only one QMP connection, please see:
https://github.com/qemu/qemu/blob/master/util/qemu-sockets.c#L936

About the async event,  I have updated code to filter out `Success` message.
Currently, seems no module wants to get the QMP events, I think no need to save them and process later.